### PR TITLE
Improve cache and performance of `get_user_data_from_wp_global_styles`.

### PR DESCRIPTION
**Pull request by spacedmonkey**
## What?

Trac ticket: https://core.trac.wordpress.org/ticket/55392
Orignal PR - https://github.com/WordPress/wordpress-develop/pull/2414

### CODE REVIEW

File: lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php [hunk 1]
```diff
@@ -16,6 +16,83 @@
...
+		}
+
+		if ( $post_id > 0 && in_array( get_post_status( $post_id ), (array) $post_status_filter, true ) ) {
+			return get_post( $post_id, ARRAY_A );
+		}
+
+		$gs_query     = new WP_Query();
+		$recent_posts = $gs_query->query( $args );
+		if ( count( $recent_posts ) === 1 ) {
+			$user_cpt = get_post( $recent_posts[0], ARRAY_A );
```
Comment (spacedmonkey) [inline lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php:72]: This fixes an issue, where a WP_Post object is returned.

File: lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php [hunk 2]
```diff
@@ -16,6 +16,83 @@
...
+					'post_title'   => __( 'Custom Styles', 'default' ),
+					'post_type'    => $post_type_filter,
+					'post_name'    => 'wp-global-styles-' . urlencode( wp_get_theme()->get_stylesheet() ),
+					'tax_input'    => array(
+						'wp_theme' => array( wp_get_theme()->get_stylesheet() ),
+					),
+				),
+				true
+			);
+			if ( ! is_wp_error( $cpt_post_id ) ) {
```
Comment (spacedmonkey) [inline lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php:87]: This check should always have been here. WP_Insert_post can sometimes return a WP_Error.

File: lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php [hunk 3]
```diff
@@ -16,6 +16,83 @@
...
+					'field'    => 'name',
+					'terms'    => $theme->get_stylesheet(),
+				),
+			),
+			'ignore_sticky_posts' => true,
+			'no_found_rows'       => true,
+		);
+
+		$cache_key = sprintf( 'wp_global_styles_%s', md5( serialize( $args ) ) );
+		$post_id   = (int) get_transient( $cache_key );
```
Comment (spacedmonkey) [inline lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php:59]: Use transient is better for performance here. Most sites do not run object caches. Transinets save to options table and will have 3 database queries per page load.

File: lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php [hunk 4]
```diff
@@ -16,6 +16,83 @@
...
+		// Special case: '-1' is a results not found.
+		if ( -1 === $post_id && ! $create_post ) {
+			return $user_cpt;
+		}
+
+		if ( $post_id > 0 && in_array( get_post_status( $post_id ), (array) $post_status_filter, true ) ) {
+			return get_post( $post_id, ARRAY_A );
+		}
```
Comment (spacedmonkey) [inline lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php:67]: Improve handling of post status changed. If this post is deleted or changed to draft.

File: lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php (hunk 2 above)
Comment (draganescu) [reply lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php:87]: What happens if `is_wp_error` is true? It seems it remains empty array and we count on array with length zero to be evaluated to false. Could we make this clear in code?

File: lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php [hunk 5]
```diff
@@ -16,6 +16,83 @@
...
+		if ( count( $recent_posts ) === 1 ) {
+			$user_cpt = get_post( $recent_posts[0], ARRAY_A );
+		} elseif ( $create_post ) {
+			$cpt_post_id = wp_insert_post(
+				array(
+					'post_content' => '{"version": ' . WP_Theme_JSON_Gutenberg::LATEST_SCHEMA . ', "isGlobalStylesUserThemeJSON": true }',
+					'post_status'  => 'publish',
+					'post_title'   => __( 'Custom Styles', 'default' ),
+					'post_type'    => $post_type_filter,
+					'post_name'    => 'wp-global-styles-' . urlencode( wp_get_theme()->get_stylesheet() ),
```
Comment (draganescu) [inline lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php:80]: The addition of `sprintf` from https://github.com/WordPress/wordpress-develop/pull/2414/files#diff-d6b86476eed058e7cf8b6e57fa52c4fd75b1f0907e1a9ccb0149528a24f7578bR298 seems legit. Shouldn't this be a copy of that other PR?

**Comment by spacedmonkey** [comment]
Love to get this into WP 6.1 cc @draganescu

**Comment by peterwilsoncc** [comment]
@spacedmonkey Now that `WP_Query` has caching, is this still required and/or as urgent?

**Comment by spacedmonkey** [comment]
Closed in favour of https://github.com/WordPress/gutenberg/pull/44290

**Pull request closed without merging**